icu4c/source/common/locid.cpp

```diff
@@ -2071,8 +2069,11 @@ Locale::initBaseName(UErrorCode &status) {
                 status = U_MEMORY_ALLOCATION_ERROR;
                 return;
             }
-            copy->fullName.copyFrom(fullName, status);
-            if (U_FAILURE(status)) { return; }
+            copy->fullName = fullName;
+            if (copy->fullName.isEmpty()) {
+                status = U_MEMORY_ALLOCATION_ERROR;
+                return;
```

Comment:
Some existing and new code calls `setToBogus()` when something goes wrong, other places don't. Please check the call sites (maybe they consistently setToBogus() after calling initBaseName()?) and make this function consistent.

Reply:
I've checked and it's consistent as far as I can see: whether `setToBogus()` is called or not after an operation fails depends on whether the failure could have left the object in a bogus state. Like on this line, failure here would leave a fully working object, just with erroneously empty `fullName`, so not bogus.

Reply:
Well, that's the way it currently is. We could change it so that any failure while modifying the object always sets it to bogus, but that'd be out-of-scope for this PR.

Reply:
> Well, that's the way it currently is.

Fair enough --> ICU-23178